dbms/src/Functions/FunctionsRegexp.h

```diff
@@ -973,15 +1114,373 @@ class FunctionStringRegexp : public FunctionStringRegexpBase
+        GetIntFuncPointerType get_pos_func = getGetIntFuncPointer(pos_param.getIntType());
+        GetIntFuncPointerType get_occur_func = getGetIntFuncPointer(occur_param.getIntType());
+        GetIntFuncPointerType get_ret_op_func = getGetIntFuncPointer(ret_op_param.getIntType());
+
+        // Container will not be used when parm is const
+        const void * pos_container =  pos_param.getContainer();
+        const void * occur_container =  occur_param.getContainer();
+        const void * ret_op_container =  ret_op_param.getContainer();
+
+        // Const value will not be used when param is not const
+        Int64 pos_const_val = PosT::isConst() ? pos_param.template getInt<Int64>(0) : -1;
+        Int64 occur_const_val = OccurT::isConst() ? occur_param.template getInt<Int64>(0) : -1;
+        Int64 ret_op_const_val = RetOpT::isConst() ? ret_op_param.template getInt<Int64>(0) : -1;
+
+        // Check if args are all const columns
+        if constexpr (ExprT::isConst() && PatT::isConst() && PosT::isConst() && OccurT::isConst() && RetOpT::isConst() && MatchTypeT::isConst())
+        {
+            if (col_size == 0 || expr_param.isNullAt(0) || pat_param.isNullAt(0) || pos_param.isNullAt(0) || occur_param.isNullAt(0) || ret_op_param.isNullAt(0) || match_type_param.isNullAt(0))
+            {
+                res_arg.column = res_arg.type->createColumnConst(col_size, Null());
+                return;
+            }
+            
+            int flags = getDefaultFlags();
+            String expr = expr_param.getString(0);
+            String pat = pat_param.getString(0);
+            if (unlikely(pat.empty()))
+                throw Exception(EMPTY_PAT_ERR_MSG);
+
+            Int64 pos = PosT::isConst() ? pos_const_val : get_pos_func(pos_container, 0);
+            Int64 occur = OccurT::isConst() ? occur_const_val : get_occur_func(occur_container, 0);
+            Int64 ret_op = RetOpT::isConst() ? ret_op_const_val : get_ret_op_func(ret_op_container, 0);
+            String match_type = match_type_param.getString(0);
+
+            Regexps::Regexp regexp(addMatchTypeForPattern(pat, match_type, collator), flags);
+            ResultType res = regexp.instr(expr.c_str(), expr.size(), pos, occur, ret_op);
+            res_arg.column = res_arg.type->createColumnConst(col_size, toField(res));
+            return;
+        }
+
+        // Initialize result column
+        auto col_res = ColumnVector<ResultType>::create();
+        typename ColumnVector<ResultType>::Container & vec_res = col_res->getData();
+        vec_res.resize(col_size, 0);
+
+        constexpr bool has_nullable_col = ExprT::isNullableCol() || PatT::isNullableCol() || PosT::isNullableCol() || OccurT::isNullableCol() || RetOpT::isNullableCol() || MatchTypeT::isNullableCol();
+
+#define GET_POS_VALUE(idx) \
+    do \
+    { \
+        if constexpr (PosT::isConst()) \
+            pos = pos_const_val; \
+        else \
+            pos = get_pos_func(pos_container, idx); \
```

Comment:
> Why not hidden `isConst` inside `get_pos_func`? So we don't need these macros

Because compiler can't inline the function as we use function pointer and we will always get a const value by calling function if we hide `isCOnst` inside `get_pos_func`.